tokio-util/src/codec/length_delimited.rs

```diff
@@ -386,8 +386,15 @@ use std::{cmp, fmt, mem};
 /// `Builder` enables constructing configured length delimited codecs. Note
 /// that not all configuration settings apply to both encoding and decoding. See
 /// the documentation for specific methods for more detail.
+///
+/// Note that the if the value of [`Builder::max_frame_length`] becomes larger than
+/// what can actually fit in [`Builder::length_field_length`], it will be clipped to
+/// the maximum value that can fit.
 #[derive(Debug, Clone, Copy)]
 pub struct Builder {
+    // Indicates whether the value of `max_frame_len` is specified by the user or is adjusted
+    max_frame_len_adjusted: bool,
```

Comment:
What is the purpose of this field?

Reply:
To distinguish between when the value of `max_frame_len` is set by the user (which we are not allowed to override with a larger value), or when the value is the result of a previous adjustment.

For example if user sets `max_frame_len` to `255`, and then `length_adjustment`, or `length_field_length` is called, we don't want to override this value even if those functions now allow a larger maximum.

I added this comment to explain when do we update the value:
```rust
        // We update the value of `max_frame_len` in two cases
        // - Current value does not fit in `length_field_len` bytes
        // - We have a larger `max_allowed_len` but only if the value has been adjusted before.
        //   This is to prevent overriding `max_frame_len` provided by the user.
        if self.max_frame_len > max_allowed_len || (adjust_to_max && self.max_frame_len_adjusted) {
            self.max_frame_len = max_allowed_len;
            self.max_frame_len_adjusted = true;
        }
```

But maybe `adjust_to_max` is not needed...

Reply:
I think you can avoid this by not adjusting the maximum until you build the codec. That way, you never need to increase it.

Reply:
Yeah that would simplify things, but `Builder::new_codec` takes `&self`, so I can't change the value then.

Reply:
```rs
pub fn new_codec(&self) -> LengthDelimitedCodec {
    let mut builder = *self;
    fixup_max_frame_len(&mut builder);
    LengthDelimitedCodec {
        builder,
        state: DecodeState::Head,
    }
}
```

Reply:
Ah I have to fix my tunnel vision :)